### ISSUE TITLE
Invalid index access on Record type

### ISSUE THREAD

**Issue by Cammisuli** [microsoft/TypeScript#45446]
# Bug Report

### 🔎 Search Terms

index access on record
index record
index type with {[key:string]: any}

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Records/Objects and accessing the properties with an index

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?strictNullChecks=false&noUncheckedIndexedAccess=true&allowUnreachableCode=true&allowUnusedLabels=true&noUnusedLocals=true&noUnusedParameters=true&noFallthroughCasesInSwitch=true&noPropertyAccessFromIndexSignature=true&noImplicitOverride=true&exactOptionalPropertyTypes=true&useUnknownInCatchVariables=true&ts=4.5.0-dev.20210813#code/MYewdgzgLgBFBcMBKBTUAnAJgHmuglmAOYA0MA3jAA7ohWJgCuAtgEYrowC+AfDALwUAUDFEwAhonIixsmnUQBGGaK4kVMVlI1za9GACYNXISdFCoAbQAMAXQB08qkIuWA5OLcOnQA)

### 💻 Code

```ts
const t: Record<string, { prop: number }> = {
    a: {
        prop: 1
    },
    b: {
        prop: 2
    }
}
  
t[0].prop

t['a'].prop
```
This also shows the issue as well. 
```ts
const t: { [key: string]: { prop: number } } = {
    a: {
        prop: 1
    },
    b: {
        prop: 2
    }
}
  
t[0].prop

t['a'].prop
```

### 🙁 Actual behavior
doing `t[0]` does not trigger a compiler error. Especially when we define the Record to have the keys as a string.

### 🙂 Expected behavior
Accessing objects where we define the key to be of a type string, shouldn't allow accessing the index with a number.

**Comment by RyanCavanaugh** [comment]
`t[0]` is the same as `t["0"]`, which is a predictable coercion.
